### ISSUE TITLE
exponentials

### ISSUE THREAD

**Issue by swight-prc** [passalis/bof_eeg#1]
Hello!

I read your paper for EURASIP 2017, and I found this repo by searching your name.

I have a question about the paper and how it's implemented here. Specifically, I'm wondering where you're doing the exponentiation.

It looks like most of equation 2 happens in [sym_histogram](https://github.com/passalis/bof_eeg/blob/20d8e2a098c073159b4bfa77e41639160b1cd639/models/neural_bof.py#L153-L162) -> [sym_distance_matrix](https://github.com/passalis/bof_eeg/blob/20d8e2a098c073159b4bfa77e41639160b1cd639/models/neural_bof.py#L188-L205). 

But I don't see the exponentiation. Is it happening somewhere and I've missed it, or am I looking at the wrong repository?

Thanks!

**Comment by passalis** [comment]
Hi! Thanks for your interest in our work.

The exponentiation is indeed implemented in [sym_histogram()](https://github.com/passalis/bof_eeg/blob/20d8e2a098c073159b4bfa77e41639160b1cd639/models/neural_bof.py#L160). However, as you correctly note, there is indeed no direct call to the exp() function. Instead, we are using [softmax()](https://github.com/passalis/bof_eeg/blob/20d8e2a098c073159b4bfa77e41639160b1cd639/models/neural_bof.py#L160), to exploit the fact that we are going then normalize with the resulting values. 

![image](https://user-images.githubusercontent.com/16520105/115833267-8279ec00-a41c-11eb-8871-f21d88477a24.png)
(image from https://en.wikipedia.org/wiki/Softmax_function)

We follow this approach, since there can be numerical issues when calculating this normalized quantity. So, here we rely on the already numerically stable implementation of softmax provided by theano.

Furthermore, you can also take a look on a subsequent work, where we can replace the Gaussian kernel used for calculating the similarities with any other kernel and we discuss options to make this process much more stable for use in deep neural networks:

N. Passalis, A. Tefas, J. Kanniainen, M. Gabbouj and A. Iosifidis, Temporal logistic neural Bag-of-Features for financial time series forecasting leveraging limit order book data, Pattern Recognition Letters, , vol. 136, pp 183-189, 2020

You can find a preprint [here](https://arxiv.org/abs/1901.08280).

**Comment by swight-prc** [comment]
I don't think I ever realized that Softmax was anything more than Current/Total. And I will definitely take a look at your new work.

Thanks!
